main.go

```diff
@@ -81,80 +114,101 @@ func run(ctx context.Context, castaiclient castai.Client, baseLogger *logrus.Log
 	log.Data["provider"] = provider.Name()
 	log.Infof("using provider %q", provider.Name())
 
-	clusterID := ""
-	if cfg.Static != nil {
-		clusterID = cfg.Static.ClusterID
-	}
+	leaderFunc := func(ctx context.Context) error {
+		clusterID := ""
+		if cfg.Static != nil {
+			clusterID = cfg.Static.ClusterID
+		}
 
-	if clusterID == "" {
-		reg, err := provider.RegisterCluster(ctx, castaiclient)
-		if err != nil {
-			return fmt.Errorf("registering cluster: %w", err)
+		if clusterID == "" {
+			reg, err := provider.RegisterCluster(ctx, castaiclient)
+			if err != nil {
+				return fmt.Errorf("registering cluster: %w", err)
+			}
+			clusterID = reg.ClusterID
+			clusterIDChanged(clusterID)
+			log.Infof("cluster registered: %v, clusterID: %s", reg, clusterID)
+		} else {
+			clusterIDChanged(clusterID)
+			log.Infof("clusterID: %s provided by env variable", clusterID)
 		}
-		clusterID = reg.ClusterID
-		log.Infof("cluster registered: %v, clusterID: %s", reg, clusterID)
-	} else {
-		log.Infof("clusterID: %s provided by env variable", clusterID)
-	}
-	castailog.SetupLogExporter(baseLogger, localLog, castaiclient, castailog.Config{
-		ClusterID:   clusterID,
-		SendTimeout: LogExporterSendTimeout,
-	})
 
-	log.Data["cluster_id"] = clusterID
+		w := &controller.Worker{
+			Fn: controller.RunController(log, clientset, castaiclient, provider, clusterID, cfg, agentVersion, ctrlHealthz),
+		}
+		defer w.Stop(log)
 
-	exitCh := make(chan error)
+		if err := w.Start(ctx); err != nil {
+			return fmt.Errorf("controller loop error: %w", err)
+		}
 
-	if cfg.PprofPort != 0 {
-		addr := fmt.Sprintf(":%d", cfg.PprofPort)
-		pprofSrv := &http.Server{Addr: addr, Handler: http.DefaultServeMux}
-		defer func() {
-			if err := pprofSrv.Close(); err != nil {
-				log.Errorf("closing pprof server: %v", err)
+		return nil
+	}
+
+	replicas.Run(ctx, log, cfg.LeaderElection, clientset, leaderWatchDog, func(ctx context.Context) {
+		exitCh <- leaderFunc(ctx)
+	})
+	return nil
+}
+
+// if any errors are observed on exitCh, context cancel is called, and all errors in the channel are logged
+func watchExitErrors(ctx context.Context, log *logrus.Entry, exitCh chan error, ctxCancel func()) {
+	select {
+	case err := <-exitCh:
+		if err != nil && !errors.Is(err, context.Canceled) {
+			log.Errorf("agent stopped with an error: %v", err)
+		}
+		ctxCancel()
+		for err := range exitCh {
```

Comment:
This loop will hang forever because the `exitCh` is never closed.

Reply:
Dunno, is there any harm for a termination handler? cure here will be worse than a disease, we'll need to be aware of all the services that might still wait to push to the channel.

Previously we had opposite situation, all services pushing to errors channel, but it's read only once, so they hang forever until process is closed.